internal/configs/virtualserver.go

```diff
@@ -490,169 +546,212 @@ type policiesCfg struct {
 	ErrorReturn     *version2.Return
 }
 
-// TODO refactor generatePolicies
-func (vsc *virtualServerConfigurator) generatePolicies(owner runtime.Object, ownerNamespace string, vsNamespace string,
-	vsName string, policyRefs []conf_v1.PolicyReference, policies map[string]*conf_v1alpha1.Policy, jwtKeys map[string]string, ingressMTLSPemFileName string, context string, tlsPemFileName string, egressMTLSSecrets map[string]string) policiesCfg {
-	var policyErrorReturn *version2.Return
-	var allow, deny []string
-	var limitReqOptions version2.LimitReqOptions
-	var limitReqZones []version2.LimitReqZone
-	var limitReqs []version2.LimitReq
-	var JWTAuth *version2.JWTAuth
-	var ingressMTLS *version2.IngressMTLS
-	var egressMTLS *version2.EgressMTLS
-	var policyError bool
+func newPoliciesConfig() *policiesCfg {
+	return &policiesCfg{}
+}
 
-	for _, p := range policyRefs {
-		polNamespace := p.Namespace
-		if polNamespace == "" {
-			polNamespace = ownerNamespace
+type policyOwnerDetails struct {
+	owner          runtime.Object
+	ownerNamespace string
+	vsNamespace    string
+	vsName         string
+}
+
+type policyOptions struct {
+	jwtKeys                map[string]string
+	ingressMTLSPemFileName string
+	tlsPemFileName         string
+	egressMTLSSecrets      map[string]string
+}
+
+type validationResults struct {
+	isError  bool
+	warnings []string
+}
+
+func newValidationResults() *validationResults {
+	return &validationResults{}
+}
+
+func (v *validationResults) addWarningf(msgFmt string, args ...interface{}) {
+	v.warnings = append(v.warnings, fmt.Sprintf(msgFmt, args...))
+}
+
+func (p *policiesCfg) addAccessControlConfig(accessControl *conf_v1alpha1.AccessControl) *validationResults {
+	res := newValidationResults()
+	p.Allow = append(p.Allow, accessControl.Allow...)
+	p.Deny = append(p.Deny, accessControl.Deny...)
+	if len(p.Allow) > 0 && len(p.Deny) > 0 {
+		res.addWarningf("AccessControl policy (or policies) with deny rules is overridden by policy (or policies) with allow rules")
+	}
+	return res
+}
+
+func (p *policiesCfg) addRateLimitConfig(rateLimit *conf_v1alpha1.RateLimit, polKey string, polNamespace string, polName string, vsNamespace string, vsName string) *validationResults {
+	res := newValidationResults()
+	rlZoneName := fmt.Sprintf("pol_rl_%v_%v_%v_%v", polNamespace, polName, vsNamespace, vsName)
+	p.LimitReqs = append(p.LimitReqs, generateLimitReq(rlZoneName, rateLimit))
+	p.LimitReqZones = append(p.LimitReqZones, generateLimitReqZone(rlZoneName, rateLimit))
+	if len(p.LimitReqs) == 1 {
+		p.LimitReqOptions = generateLimitReqOptions(rateLimit)
+	} else {
+		curOptions := generateLimitReqOptions(rateLimit)
+		if curOptions.DryRun != p.LimitReqOptions.DryRun {
+			res.addWarningf("RateLimit policy %q with limit request option dryRun=%v is overridden to dryRun=%v by the first policy reference in this context", polKey, curOptions.DryRun, p.LimitReqOptions.DryRun)
```

Comment:
Optional suggestion: 
* Split this line over multiple lines to reduce line length

```suggestion
			res.addWarningf(
				"RateLimit policy %q with limit request option dryRun=%v is overridden to dryRun=%v by the first policy reference in this context",
				polKey, curOptions.DryRun, p.LimitReqOptions.DryRun,
			)
```

Reply:
Thanks, I really wasn't sure what the practice is in go to split long lines. 
I figured if `gofmt` didn't split it then I should leave it alone, but some googling found this tool: https://github.com/segmentio/golines
Any idea if the line-splitting recommended there is what's usually done in go code?

Reply:
There's not actually a standard and it's up to personal preference I guess. We could start using golines and agree on a line length. I'd vote for 120 😄

Reply:
Maybe: `golangci-lint run --enable lll`

`lll` will enable line length at 120 and can easily be added to our pipelines.
(https://github.com/nginxinc/kubernetes-ingress/blob/master/.golangci.yml)

Ref:
https://github.com/walle/lll

Reply:
I used golines locally on the files in this PR to limit length to 120 chars.
I think it's a great suggestion to start linting for this in the pipeline, or incorporate golines into the pipeline somehow in a future PR.